fix: [Apple TV] nextFocus props is not correct while set for the focused component

dratwas opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

In this PR I added fix for an issue with `nextFocusXXX` props that are set for the already focused component.

Native FocusGuides are added to the view while the item is focused and removed while blurred. If the `nextFocusXXX` is added to an already focused item the value is not updated (or the focus guide is not added if the previous value was not set)

In my solution, I split the `enableDirectionalFocusGuides` into a few smaller methods and used them in `nextFocusXXX` prop setter.

I check if the view in which we set the property is focused and if yes I refresh the suitable focusGuide environment.

## Changelog:

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:


For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

```
[iOS][Fixed] - Update focus guide environment while setting nextFocus prop for already focused component
```

## Test Plan:

Change the `nextFocusXXX` prop (nextFocusRight on the video) when the item is focused and check if the destination view is correct.

**Video before**

https://github.com/react-native-tvos/react-native-tvos/assets/16336501/322ac15f-8d56-4035-a887-87f95fe2e23b

**Video after**


https://github.com/react-native-tvos/react-native-tvos/assets/16336501/1b82d57d-aa2b-4c22-8a2c-dc775910c0a3



<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

<details>
<summary><b>Code snippet that can be pasted in examples</b></summary>

```
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow
 */

'use strict';

const React = require('react');
const ReactNative = require('react-native');

import {useRNTesterTheme} from '../../components/RNTesterTheme';

const {View, StyleSheet, Pressable, Text, TVFocusGuideView} = ReactNative;

const screenHeight = ReactNative.Dimensions.get('window').height;
const scale = screenHeight / 1080;
const width = 200 * scale;
const height = 120 * scale;

exports.framework = 'React';
exports.title = 'TV NextFocus example';
exports.description = 'NextFocus API';
exports.displayName = 'NextFocus example';
exports.examples = [
  {
    title: 'TVFocusGuide',
    render(): React.Node {
      return <TVFocusGuideExample />;
    },
  },
];

const Button = React.forwardRef((props: $FlowFixMeProps, _) => {
  const onRefAssign = ref => {
    props.onRefAssign?.(props.index, ref);
  };
  const theme = useRNTesterTheme();
  return (
    <Pressable
      onPress={props.onPress}
      nextFocusRight={props.nextFocusRight}
      style={({pressed, focused}) =>
        focused ? styles.buttonStyleFocused : styles.buttonStyle
      }
      ref={onRefAssign}>
      <Text style={[{color: theme.LinkColor}, styles.buttonText]}>
        {props.label}
      </Text>
    </Pressable>
  );
});

const TVFocusGuideExample = () => {
  const [nextFocusIndex, setNextFocusIndex] = React.useState(0);

  const refs = React.useRef({});
  const onRefAssign = (index, ref) => {
    refs.current[index] = ref;
  };

  return (
    <View style={styles.rowContainer}>
      <TVFocusGuideView autoFocus>
        <Button
          label={`Focus to ${nextFocusIndex}`}
          nextFocusRight={refs.current[nextFocusIndex]}
          onPress={() => {
            setNextFocusIndex((nextFocusIndex + 1) % 4);
          }}
        />
      </TVFocusGuideView>
      <View>
        <Button label={'Item nr 0'} onRefAssign={onRefAssign} index={0} />
        <Button label={'Item nr 1'} onRefAssign={onRefAssign} index={1} />
        <Button label={'Item nr 2'} onRefAssign={onRefAssign} index={2} />
        <Button label={'Item nr 3'} onRefAssign={onRefAssign} index={3} />
      </View>
    </View>
  );
};

const marginSize = 20 * scale;
const styles = StyleSheet.create({
  rowContainer: {
    flexDirection: 'row',
    padding: 100 * scale,
  },
  buttonText: {
    fontSize: 30 * scale,
  },
  buttonStyle: {
    marginLeft: marginSize,
    marginRight: marginSize,
    marginTop: marginSize,
    marginBottom: marginSize,
  },
  buttonStyleFocused: {
    opacity: 0.5,
    width,
    height,
    marginLeft: marginSize,
    marginRight: marginSize,
    marginTop: marginSize,
    marginBottom: marginSize,
  },
});

```

</details>

yusufyildirim commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -432,6 +432,17 @@ - (BOOL)canBecomeFocused
 // This is only done once the view is focused.
 //
 - (void)enableDirectionalFocusGuides
+{
+  [self enableNextFocusUp];
+
+  [self enableNextFocusDown];
+
+  [self enableNextFocusLeft];
+
+  [self enableNextFocusRight];
```
I'd personally delegate the responsibility of checking if the view is focused or not to this function. Effectively making it defensive. It'd protect us from accidentally calling it without the check in some other places

douglowder replied:
Agreed

yusufyildirim commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -969,6 +998,9 @@ - (void)updateProps:(const Props::Shared &)props oldProps:(const Props::Shared &
     if (newViewProps.nextFocusRight.has_value()) {
       UIView *rootView = [self containingRootView];
       _nextFocusRight = [rootView viewWithTag:newViewProps.nextFocusRight.value()];
+      if (self.isFocused) {
+        [self enableNextFocusRight];
+      }
```
I'd just call the `enableDirectionalFocusGuides` wherever I need it combined with the way I described above. So I wouldn't create separate `enableNext*` functions. You'd make this PR quite a bit shorter that way.

yusufyildirim reviewed (commented):
Thanks for the PR! I left some comments but I'm also totally ok with the current implementation. Leaving the decision to @douglowder. One question, though. There's no Android change, is that how it already works?

douglowder reviewed (changes requested):
Looks good! Before merging, I would like to see the changes requested by @yusufyildirim in [this comment](https://github.com/react-native-tvos/react-native-tvos/pull/658/files#r1486363985)

dratwas commented:
Hey @yusufyildirim and @douglowder thanks for the review! I applied the requested changes and added an example. I also added a video for Android. As mentioned in description - Android works fine w/o any changes ;)
